pkg/gcp/gcp.go

```diff
@@ -0,0 +1,178 @@
+}
+
+// NewCloud creates a new api.Cloud instance which can prepare GCP for Submariner to be deployed on it
+func NewCloud(infraID string, client gcpclient.Interface) api.Cloud {
+	return &gcpCloud{
+		infraID:   infraID,
+		projectID: client.GetProjectID(),
+		client:    client,
+	}
+}
+
+// PrepareForSubmariner prepares submariner cluster environment on GCP
+func (gc *gcpCloud) PrepareForSubmariner(input api.PrepareForSubmarinerInput, reporter api.Reporter) error {
+	// create the inbound and outbound firewall rules for submariner public ports
+	ingress, egress := gc.newFirewallRules(publicPortsRuleName, input.PublicPorts, false)
+	if err := gc.openPorts(ingress, egress); err != nil {
+		return fmt.Errorf("failed to open submariner public ports: %v", err)
+	}
+
+	reporter.Succeeded("Submariner public ports are opened with firewall rules %q and %q on GCP", ingress.Name, egress.Name)
+
+	// create the inbound firewall rule for submariner internal ports
+	internalIngress, _ := gc.newFirewallRules(internalPortsRuleName, input.InternalPorts, true)
+	if err := gc.openPorts(internalIngress); err != nil {
+		return fmt.Errorf("failed to open submariner internal ports: %v", err)
+	}
+
+	reporter.Succeeded("Submariner internal ports are opened with firewall rule %q on GCP", internalIngress.Name)
+
+	return nil
+}
+
+// CleanupAfterSubmariner clean up submariner cluster environment on GCP
+func (gc *gcpCloud) CleanupAfterSubmariner(reporter api.Reporter) error {
+	ingressName, egressName := generateRuleNames(gc.infraID, publicPortsRuleName)
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	if err := gc.client.DeleteFirewallRule(ingressName); err != nil {
+		return err
+	}
+
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner public ports are closed on GCP")
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(internalIngressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner internal ports are closed on GCP")
+
+	return nil
+}
+
+func (gc *gcpCloud) newFirewallRules(name string, ports []api.PortSpec, isInternal bool) (ingress, egress *compute.Firewall) {
+	ingressName, egressName := generateRuleNames(gc.infraID, name)
+
+	allowedPorts := []*compute.FirewallAllowed{}
+	for _, port := range ports {
+		allowedPorts = append(allowedPorts, &compute.FirewallAllowed{
+			IPProtocol: port.Protocol,
+			Ports:      []string{strconv.Itoa(int(port.Port))},
+		})
+	}
+
+	ingress = &compute.Firewall{
+		Name:      ingressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "INGRESS",
+		Allowed:   allowedPorts,
+	}
+
+	// if ports are internal, we make the ports are accessed in internal cluster and egress is not required
```

Comment:
Please split to 2 functions, one for external and one for internal. It'll be cleaner and simpler to understand.

Something like:
```go
func (gc *gcpCloud) newFirewallRule(name, direction string, ports []api.PortSpec) *compute.Firewall {
	allowedPorts := []*compute.FirewallAllowed{}
	for _, port := range ports {
		allowedPorts = append(allowedPorts, &compute.FirewallAllowed{
			IPProtocol: port.Protocol,
			Ports:      []string{strconv.Itoa(int(port.Port))},
		})
	}

	return &compute.Firewall{
		Name:      name,
		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
		Direction: direction,
		Allowed:   allowedPorts,
	}
}

func (gc *gcpCloud) newInternalFirewallRule(name string, ports []api.PortSpec) *compute.Firewall {
	ingressName, _ := generateRuleNames(gc.infraID, name)

	rule := gc.newFirewallRule(ingressName, "INGRESS", ports)
	rule.TargetTags = []string{
		fmt.Sprintf("%s-worker", gc.infraID),
		fmt.Sprintf("%s-master", gc.infraID),
	}
	rule.SourceTags = []string{
		fmt.Sprintf("%s-worker", gc.infraID),
		fmt.Sprintf("%s-master", gc.infraID),
	}

	return rule
}

func (gc *gcpCloud) newExternalFirewallRules(name string, ports []api.PortSpec) (ingress, egress *compute.Firewall) {
	ingressName, egressName := generateRuleNames(gc.infraID, name)

	return gc.newFirewallRule(ingressName, "INGRESS", ports), gc.newFirewallRule(egressName, "EGRESS", ports)
}
```

Reply:
separated